### ISSUE TITLE
Compiler bug: "expected an instance declaration"

### ISSUE THREAD

**Issue by jnfoster** [p4lang/p4c#807]
While trying to write a devious cyclic example for @jafingerhut, I discovered a new (to me) compiler bug in the inlining pass of the midend:
```
#include <core.p4>

// Architecture    
control C1();
control C2();
control C3();
extern E {
  E();
  void set1(C1 c1);
  void set2(C2 c2);
  C1 get1();
  C2 get2();
}

package S(C3 c3);

// User Program            
control MyC1()(E e) {
  apply {
    C2 c2 = e.get2();
    c2.apply();
  }
}        
control MyC2()(E e) {
  apply {
    C1 c1 = e.get1();
    c1.apply();
  }
}        
control MyC3()(E e, C1 c1, C2 c2) {
  apply {
    e.set1(c1);
    e.set2(c2);
    c1.apply();
  }
}        

E() e;
    
S(MyC3(e,MyC1(e),MyC2(e))) main;
```
Error:
```
In file: p4c/midend/inlining.cpp:404
Compiler Bug: cycle.p4(18): C2 c2_0/c2 expected an instance declaration
    C2 c2 = e.get2();
    ^^^^^^^^^^^^^^^^^
```

**Comment by mbudiu-vmw** [comment]
I think that we'll fix this one by forbidding methods to return controls (parsers, packages, or other externs).

**Comment by jnfoster** [comment]
The following program produces the same error message:
```
#include <core.p4>

// Architecture    
control C();
package S(C c);
extern BoolRegister {
  BoolRegister();
  bool get();
  void flip();
}        

// User Program
control MyC1() {
  apply {
  }
}
control MyC2() {
  apply {
  }
}        
control MyC3()(BoolRegister r, C c1, C c2) {
  apply {
    r.flip();
    if(r.get()) c1.apply();
    else c2.apply();
  }
}
control MyC4() {
  BoolRegister() r;
  MyC1() c1;
  MyC2() c2;
  MyC3(r,c1,c2) c3;
  apply {
    c3.apply();
  }  
}        
S(MyC4()) main;
```

```
In file: p4c/midend/inlining.cpp:404
Compiler Bug: cyclic.p4(21):  C c1 expected an instance declaration
control MyC3()(BoolRegister r, C c1, C c2) {
                                 ^^
```

It only uses controls as constructor parameters, which is definitely legal. It also seems unlikely this is fixed by 7317e5fc5bb4d56761287cbc86be092aafadf322?
